llama-index-integrations/agent/llama-index-agent-introspective/llama_index/agent/introspective/reflective/self_reflection.py

```diff
@@ -0,0 +1,430 @@
+
+
+class SelfReflectionAgentWorker(BaseModel, BaseAgentWorker):
+    """Self Reflection Agent Worker.
+
+    This agent performs a reflection without any tools on a given response
+    and subsequently performs correction. It should be noted that this reflection
+    implementation has been inspired by two works:
+
+    1. Reflexion: Language Agents with Verbal Reinforcement Learning, by Shinn et al. (2023)
+        (https://arxiv.org/pdf/2303.11366.pdf)
+    2. CRITIC: Large Language Models Can Self-Correct with Tool-Interactive Critiquing, by Gou et al. (2024)
+       (https://arxiv.org/pdf/2305.11738.pdf)
+
+    This agent performs cycles of reflection and correction on an initial response
+    until a satisfactory correction has been generated or a max number of cycles
+    has been reached. To perform reflection, this agent utilizes a user-specified
+    LLM along with a PydanticProgram to generate a structured output that contains
+    an LLM generated reflection of the current response. After reflection, the
+    same user-specified LLM is used again but this time with another PydanticProgram
+    to generate a structured output that contains an LLM generated corrected
+    version of the current response against the priorly generated reflection.
+
+    Attr:
+        max_iterations (int, optional): The max number of reflection & correction.
+            Defaults to DEFAULT_MAX_ITERATIONS.
+        callback_manager (Optional[CallbackManager], optional): Callback manager.
+            Defaults to None.
+        llm (Optional[LLM], optional): The LLM used to perform reflection and correction.
+            Must be an OpenAI LLM at this time. Defaults to None.
+        verbose (bool, optional): Whether execution should be verbose. Defaults to False.
+    """
+
+    callback_manager: CallbackManager = Field(default=CallbackManager([]))
+    max_iterations: int = Field(default=DEFAULT_MAX_ITERATIONS)
+    _llm: LLM = PrivateAttr()
+    _reflection_program: BaseLLMFunctionProgram = PrivateAttr()
+    _correction_program: BaseLLMFunctionProgram = PrivateAttr()
+    _verbose: bool = PrivateAttr()
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(
+        self,
+        max_iterations: int = DEFAULT_MAX_ITERATIONS,
+        callback_manager: Optional[CallbackManager] = None,
+        llm: Optional[LLM] = None,
+        verbose: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        """__init__."""
+        self._llm = llm
+        self._verbose = verbose
+
+        # define _reflection and _correction programs
+        try:
+            from llama_index.program.openai import OpenAIPydanticProgram
+        except ImportError:
+            raise ImportError(
+                "Missing OpenAIPydanticProgram. Please run `pip install llama-index-program-openai`."
+            )
+        self._reflection_program = OpenAIPydanticProgram.from_defaults(
```

Comment:
Instead of using programs (and locking to `OpenAIPydanticProgram` -- we can use `llm.structured_predict()` which picks the program that matches the LLM being used

Reply:
Oh cool. Let me try using that instead for both reflection classes. Less params for a user to have to supply then. :)

Reply:
this is done!